// @strapi/generate-new/dist/resources/json/common/package.json.js
"use strict";
const path = require("path");
const _ = require("lodash");
const fse = require("fs-extra");
const engines = require("./engines.js");
const _interopDefault = (e) => e && e.__esModule ? e : { default: e };
const fse__default = /* @__PURE__ */ _interopDefault(fse);
const createPackageJSON = async (scope) => {
  const pkg = {
    name: _.kebabCase(scope.name),
    private: true,
    version: "0.1.0",
    description: "A Strapi application",
    scripts: {
      develop: "strapi develop",
      start: "strapi start",
      build: "strapi build",
      strapi: "strapi",
      deploy: "strapi deploy"
    },
    devDependencies: scope.devDependencies,
    dependencies: scope.dependencies,
    author: {
      name: "A Strapi developer"
    },
    strapi: {
      ...scope.packageJsonStrapi,
      uuid: scope.uuid
    },
    engines,
    license: "MIT"
  };
  await fse__default.default.writeJSON(path.join(scope.rootPath, "package.json"), pkg, {
    spaces: 2
  });
};
module.exports = createPackageJSON;
